Improved tooltips, sanitized name, and synchronizing 2.8 fixes

TheDuckCow opened the pull request:
- Adds the `poselib.library_name_sanitize` operator, which somewhat intelligently takes the active post library name and updates to ensure it appears in the library dropdown (screenshot below)
- Improves tooltips to be more verbose and matching the default user preferences
- Synchronizing some changes from the forked 2.8 repo (https://github.com/jasperges/pose-thumbnails/commit/cedcbdfec7c9ccb88e055ef9c24e5ec528664231)

![sanitize library name button](https://user-images.githubusercontent.com/2958461/46589516-51ad5380-ca78-11e8-810b-b9839c7d3bbd.png)

sybrenstuvel commented on pose_thumbnails/core.py:
```diff
@@ -525,7 +526,9 @@ def _finish(self, context):
         POSELIB_OT_mix_pose.is_running = None
-        context.area.tag_redraw()
+
+        if context is not None:
```
In which case is `_finish(None)` called?

sybrenstuvel commented on pose_thumbnails/core.py:
```diff
@@ -546,6 +549,11 @@ def _execute(self, context, auto_key: bool):
         return {'FINISHED'}
 
     def modal(self, context, event):
+        if self._target_state == 'ABORTED':
```
When is `'ABORTED'` used? Please include all relevant code changes in the pull request.

TheDuckCow replied:
Originally this pulled in to match your 2.8 branch, but then I saw that relied upon using handlers that don't exist before 2.8 (undo_pre and redo_pre) - looks like I forgot to back out these parts of that merge.

But a more important question, is there an equivalent way this should be done in 2.7X, to avoid undo/redo from trying to access invalid pointers?

Other changes largely updated locally, will push once I have more direction on this part.

sybrenstuvel commented on pose_thumbnails/core.py:
```diff
@@ -610,6 +618,41 @@ def _determine_poses(self):
 
 
+class POSELIB_OT_sanitize_library_name(bpy.types.Operator):
+    """Make the active pose library name suitable for a character"""
```
I think "sanitize" doesn't quite cover the semantics of this operator; "rename_for_character" covers it better IMO. You can then use that in the label as well, which gives you enough space in the docstring to explain how that's done (rather than explaining what "sanitize" means).

sybrenstuvel commented on pose_thumbnails/core.py:
```diff
@@ -610,6 +618,41 @@ def _determine_poses(self):
+    def execute(self, context):
+        addon_prefs = prefs.for_addon(context)
+        if not context.object or not context.object.pose_library:
+            return {'FINISHED'}
```
`'FINISHED'` is incorrect here, as the operator didn't do anything. Use `'CANCELLED'` instead.

sybrenstuvel commented on pose_thumbnails/core.py:
```diff
@@ -610,6 +618,41 @@ def _determine_poses(self):
+        libraries = [lib[0] for lib in pose_lib_for_char_items(self, context)]
+
+        if pose_lib.name in libraries:
+            return {'FINISHED'}
```
`'FINISHED'` is incorrect here, as the operator didn't do anything. Use `'CANCELLED'` instead.

sybrenstuvel commented on pose_thumbnails/core.py:
```diff
@@ -610,6 +618,41 @@ def _determine_poses(self):
+
+        obj = context.object
+        pose_lib = context.object.pose_library
+        libraries = [lib[0] for lib in pose_lib_for_char_items(self, context)]
```
`pose_lib_for_char_items()` effectively does two things (construct `pose_libs_for_current_char`, and iterate over it to construct the result).  Here you call `pose_lib_for_char_items()` and then iterate over the result again to undo one of those two things. A better approach would be to split up `pose_lib_for_char_items()` into two functions that do one thing each, and then just call the function you need.

sybrenstuvel commented on pose_thumbnails/core.py:
```diff
@@ -610,6 +618,41 @@ def _determine_poses(self):
+
+        if pose_lib.name in libraries:
+            return {'FINISHED'}
+        else:
```
No need to use `else:` after a `return`; it unnecessarily indents the rest of the code. It also makes the core functionality of the code part of a conditional, which is undesirable.

sybrenstuvel commented on pose_thumbnails/core.py:
```diff
@@ -610,6 +618,41 @@ def _determine_poses(self):
+        if pose_lib.name in libraries:
+            return {'FINISHED'}
+        else:
+            pattern = r'(?i)(' + obj.name + r'){0,1}[._ -]'
```
This code is doing weird things. I would expect you to use the `character_name()` function to determine the character name given the rig name, and then use that to construct the library name. Right now this seems to ignore the user's configured regexp, which is a bad idea.

TheDuckCow commented:
@sybrenstuvel I believe I've addressed all your comments, including the function splitting. 

Regarding the "rename_for_character", I simplified what I was originally trying to do with regex. Essentially trying to account for these starting names and corrected updated names:
- `PLB-CharName-body_poses` -> already good, will have nothing to update
- `Plb-CharNamebody_poses` -> already good, will have nothing to update
- `body_poses` -> This now updates to `PLB-CharNamebody_poses`; In the previous commit, I was inserting a hyphen automatically between the charname and the library name, but perhaps it's better to leave that assumption to the user, even if it makes this case less immediately readable?
- `CharName-body_poses` -> Updates to `PLB-CharName-body_poses`, as opposed to the simpler logic but less desirable output of `PLB-CharNameCharName-body_poses`.
- `PLB-body_poses` -> Updates to `PLB-CharName-body_poses`, recognizing the prefix was there.
- `body_poses-CharName` -> I'm not accounting for this mis-setup, so this would still be created as `PLB-CharNamebody_poses-CharName` which personally I think is fine, the user could correct it understanding the intended pattern.

Let me know if you have any updated thoughts.

TheDuckCow commented:
@jasperges not sure if you want to weigh in here at all too, let me know if you have any supplemental comments. Your call of course to consider when to consider this mergeable. 

Personal side note on the versioning side of the house (again, just my 2c): I'm sure people have been using this addon, and in blender they see it as v1.0.0; so really, when these changes are in, would it not be appropriate to call it v1.0.1 and then just backdate adding the tag for v1.0.0 to an appropriate spot? I only say that it make it clearer for individuals (... or automated systems, in my own case) to pick up the updated changes.

jasperges commented:
@TheDuckCow Sorry, for the late response. I'm swamped with work at the moment. Will look at it as soon as possible.

jasperges commented:
@TheDuckCow So I finally made some time to look into this. I will follow your suggestion and do a v1.0.1 release (and add a v1.0.0 tag somewhere at an older commit). Unfortunately I encountered a bug when quickly testing your code (see message above). Do you have time to fix this? Then I will do the release.
By the way: in the README I added a link to the wiki (still empty). I will also update the wiki entry in the add-on itself. That way it's easy to add and update documentation.

TheDuckCow commented:
Updates made in this pull request: https://github.com/jasperges/pose-thumbnails/pull/53
Sounds good for the wiki, happy to help there that if you're strapped for time.

jasperges commented:
I don't mind at all if you make a start with the wiki. My time to work on this will be extremely limited for the foreseeable future. So any help is appreciated.
For me it's not a requirement for a new release, though. That's also why I want to use the wiki (other options are also fine with me). I want to have the documentation separate from the source code.

TheDuckCow commented:
Agreed - better to have the updated code out and tagged for 1.0.1 than delay for movement in documentation. I'll move the conversation to a doc-issue thread.
